power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -714,50 +766,113 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
+                std::all_of(all_spans.cbegin() + 1, all_spans.cend(), [&first_span](auto const& current_span) {
+                    return std::ranges::equal(
+                        current_span, first_span,
+                        [](UpdateType<CT> const& obj, UpdateType<CT> const& first) { return obj.id == first.id; });
+                });
         };
 
-        auto const is_component_update_independent = [&update_data, &process_buffer_span]<typename CT>() -> bool {
+        auto const check_each_component = [&update_data, &process_buffer_span]<typename CT>() -> UpdateCompProperties {
             // get span of all the update data
-            if (update_data.is_columnar(CT::name)) {
-                return process_buffer_span.template operator()<CT>(
-                    update_data.get_columnar_buffer_span_all_scenarios<meta_data::update_getter_s, CT>());
+            auto const comp_index = update_data.find_component(CT::name, false);
+            UpdateCompProperties result;
+            result.name = CT::name;
+            result.is_columnar = update_data.is_columnar(result.name);
+            if (comp_index >= 0) {
+                result.elements_ps_in_update = update_data.get_component_info(comp_index).elements_per_scenario;
+            }
+            if (result.is_columnar) {
+                process_buffer_span.template operator()<CT>(
+                    update_data.get_columnar_buffer_span_all_scenarios<meta_data::update_getter_s, CT>(), result);
+            } else {
+                process_buffer_span.template operator()<CT>(
+                    update_data.get_buffer_span_all_scenarios<meta_data::update_getter_s, CT>(), result);
             }
-            return process_buffer_span.template operator()<CT>(
-                update_data.get_buffer_span_all_scenarios<meta_data::update_getter_s, CT>());
+            return result;
         };
 
-        // check all components
-        auto const update_independent = run_functor_with_all_types_return_array(is_component_update_independent);
-        return std::ranges::all_of(update_independent, [](bool const is_independent) { return is_independent; });
+        // check and return indenpendence of all components
+        return run_functor_with_all_types_return_vector(check_each_component);
+    }
+
+    static bool is_update_independent(ConstDataset const& update_data) {
+        // If the batch size is (0 or) 1, then the update data for this component is 'independent'
+        if (update_data.batch_size() <= 1) {
+            return true;
+        }
+        auto const all_comp_update_independence = check_components_independence(update_data);
+        return std::ranges::all_of(all_comp_update_independence,
+                                   [](auto const& comp) { return !comp.has_id || comp.ids_match; });
+    }
+
+    void validate_update_data_independence(UpdateCompProperties const& comp) const {
+        if (!comp.has_id && comp.ids_all_na) {
```

Comment:
Btw, I really like that you encapsulated all the errors in one place :). Makes it very readable.